submission_runner.py

```diff
@@ -0,0 +1,160 @@
+from typing import Tuple
+
+from absl import app
+from absl import flags
+from absl import logging
+import collections
+import importlib
+import os
+import struct
+import time
+
+import jax
+import spec
+from workloads.mnist import workload as mnist_workload
+
+
+flags.DEFINE_string(
+    'submission_path',
+    'workloads/mnist/submission.py',
+    'The relative path of the Python file containing submission functions. '
+    'NOTE: the submission dir must have an __init__.py file!')
+flags.DEFINE_string('workload', 'mnist', 'The name of the workload to run.')
+
+FLAGS = flags.FLAGS
+
+
+# TODO(znado): make a nicer registry of workloads that lookup in.
+WORKLOADS = {
+    'mnist': mnist_workload.MnistWorkload(),
+}
+
+
+# Example reference implementation showing how to use the above functions
+# together.
+def train_once(
+    workload: spec.Workload,
+    init_optimizer_state: spec.InitOptimizerFn,
+    update_params: spec.UpdateParamsFn,
+    data_selection: spec.DataSelectionFn,
+    hyperparameters: spec.Hyperparamters,
+    rng: spec.RandomState) -> Tuple[spec.Timing, spec.Steps]:
+  data_rng, opt_init_rng, model_init_rng, rng = jax.random.split(rng, 4)
+
+  # Workload setup.
+  input_queue = workload.build_input_queue(
+      data_rng, 'train', batch_size=hyperparameters.batch_size)
+  optimizer_state = init_optimizer_state(
+      workload,
+      workload.param_shapes,
+      hyperparameters,
+      opt_init_rng)
+  model_params, model_state = workload.init_model_fn(
+      workload.param_shapes, model_init_rng)
+
+  # Bookkeeping.
+  goal_reached = False
+  is_time_remaining = True
+  last_eval_time = 0
+  accumulated_submission_time = 0
+  eval_results = []
+  global_step = 0
+  eval_now = False
+
+  while (is_time_remaining and not goal_reached):
+    step_rng = jax.random.fold_in(rng, global_step)
+    data_select_rng, preprocess_rng, update_rng, eval_rng = jax.random.split(
+        step_rng, 4)
+    start_time = time.time()
+    selected_train_input_batch, selected_train_label_batch = data_selection(
```

Comment:
Not a Jax user, so sorry if this is a silly question.
Could this same code run on a multi-gpu system, or is this specific to a single GPU? (Or is this currently CPU-only?)

Reply:
For Jax this should work on CPU/GPU/TPU!